# HG changeset patch
Pattern.unify: raise Unif only for rigid occurrences of the flexible head, otherwise Pattern

diff --git a/src/Pure/pattern.ML b/src/Pure/pattern.ML
--- a/src/Pure/pattern.ML
+++ b/src/Pure/pattern.ML
@@ -122,6 +122,10 @@
 
 fun ints_of' env ts = ints_of (map (Envir.head_norm env) ts);
 
+fun rigid_occurs(F,t,env) = case strip_comb (strip_abs_body (Envir.head_norm env t)) of
+      (Var(G,_),ts) => F=G andalso ((ints_of' env ts; true) handle Pattern => false)
+    | (_,ts) => exists (fn u => rigid_occurs(F,u,env)) ts;
+
 
 fun app (s,(i::is)) = app (s$Bound(i),is)
   | app (s,[])      = s;
@@ -267,7 +271,9 @@
      else fold (unif context binders) (ss~~ts) env
 
 and flexrigid context (params as (env,binders,F,Fty,is,t)) =
-      if occurs(F,t,env) then (ocheck_fail context (F,t,binders,env); raise Unif)
+      if occurs(F,t,env) then
+        if rigid_occurs(F,t,env) then (ocheck_fail context (F,t,binders,env); raise Unif)
+        else raise Pattern
       else (let val (u,env') = proj(t,env,binders,is)
             in Envir.update ((F, Fty), mkabs (binders, is, u)) env' end
             handle Unif => (proj_fail context params; raise Unif));
